tests/Functional/DocumentManager/QueryBuilderTest.php

```diff
@@ -0,0 +1,272 @@
+        $issue3->setName('The hot water at ten and a closed car at four');
+        $issue3->setStatus('open');
+
+        $manager->persist($issue3, null, [
+            'path' => '/test/issue3',
+            'auto_create' => true,
+        ]);
+
+        $manager->flush();
+    }
+
+    /**
+     * It should query from an alias source.
+     */
+    public function testFromAlias()
+    {
+        $manager = $this->getDocumentManager();
+        $builder = $manager->createQueryBuilder();
+        $builder->setLocale('en');
+        $query = $builder->from()->document('full', 'p')->end()->getQuery();
+        $results = $query->execute();
+        $this->assertGreaterThan(1, count($results));
+    }
+
+    /**
+     * It should query from a class FQN source.
+     */
+    public function testFromDocumentClass()
+    {
+        $manager = $this->getDocumentManager();
+        $builder = $manager->createQueryBuilder();
+        $builder->setLocale('en');
+        $query = $builder->from()->document('Sulu\Component\DocumentManager\Tests\Functional\Model\FullDocument', 'p')->end()->getQuery();
+        $results = $query->execute();
+        $this->assertGreaterThan(1, count($results));
+    }
+
+    /**
+     * Test query by field.
+     */
+    public function testQueryByField()
+    {
+        $builder = $this->getDocumentManager()->createQueryBuilder();
+        $builder
+            ->setLocale('en')
+            ->from()
+                ->document('full', 'f')
+            ->end()
+            ->where()
+                ->eq()->field('f.title')->literal('Hello');
+
+        $results = $builder->getQuery()->execute();
+        $this->assertEquals(1, count($results));
+
+        return $builder;
+    }
+
+    /**
+     * It should take the locale into account when querying.
+     *
+     * @depends testQueryByField
+     */
+    public function testQueryByFieldWithLocale($builder)
+    {
+        $builder->setLocale('az');
+
+        $results = $builder->getQuery()->execute();
+        $this->assertEquals(0, count($results));
+    }
+
+    /**
+     * It should query on fields mapped by subscribers.
+     */
+    public function testQueryBySubscriberFields()
+    {
+        $oneHourAgo = new \DateTime();
+        $oneHourAgo->modify('-1 hour');
+        $builder = $this->getDocumentManager()->createQueryBuilder();
+        $builder
+            ->setLocale('en')
+            ->from()
+                ->document('full', 'f')
+            ->end()
+            ->where()
+                ->fieldIsset('f.created');
+
+        $results = $builder->getQuery()->execute();
+        $this->assertEquals(2, count($results));
+    }
+
+    /**
+     * It should join sources, hydrating objects from right().
+     */
+    public function testQueryBuilderFromJoinHydrateRight()
+    {
+        $builder = $this->getDocumentManager()->createQueryBuilder();
+        $builder->setLocale('en');
+        $builder->from('i')->joinInner()
```

Comment:
Because `document()` is only one type of "source". The source may also be another join.

The `equi` terminology stems from PHPCR (i.e. [JCR](https://www.day.com/maven/jsr170/javadocs/jcr-2.0/javax/jcr/query/qom/EquiJoinCondition.html) and means `equi join condition`. So a join based on an equality condition.

I agree that `->eq()` would be better than `->equi()`, the ODM query builder tried to copy as far as possible the PHPCR API.

But unless we intend to write a new query builder, this is out of scope, as these are all details of the PHPCR-ODM query builder.